### DISCUSSION TITLE
Set variables as upper case for the first letter is necessary!!!

### DISCUSSION THREAD

**Discussion by KobyashiMaru** [EvgSkv/logica#156] [General]
This is my first day trying out ```Logica```, I connected to BigQuery and coded with Google Colab, and I notice something a bit weird, especially when I tried to make a query with ```limit``` and ```order by```. 

I'm not sure if this is a bug, or ```Logica``` worked intentionally like that.

When I set up a variable called "subset", as you can see, all the letters are lower cases, and it passed and worked normally
```
%%logica subset


subset(part_num:, cat_id: , name: ) :- 
  hao_test.parts(part_num:, name:, part_cat_id : cat_id), 
  cat_id == 1, 
  Like(name, "%Red%") && Like(name, "%16 x 24%")
```

However, when I set up the same variable "subset" with ```limit``` and ```order by```, it makes an error as following

```
%%logica subset

@OrderBy(subset, "cat_id desc");
@Limit(subset, 10);
subset(part_num:, cat_id:, name:) :- 
  hao_test.parts(part_num:, name:, part_cat_id: cat_id);

---------------------------------------------------------------------------

RuleCompileException                      Traceback (most recent call last)

<ipython-input-206-bfb33d84fc6d> in <module>()
----> 1 get_ipython().run_cell_magic('logica', 'subset', '\n@OrderBy(subset, "cat_id desc");\n@Limit(subset, 10);\nsubset(part_num:, cat_id:, name:) :- \n  hao_test.parts(part_num:, name:, part_cat_id: cat_id);')

12 frames

/usr/local/lib/python3.7/dist-packages/logica/compiler/universe.py in __contains__(self, key)
    388                 'variable %s.' % (
    389                     color.Warn(key)),
--> 390                 rule_text)
    391         flag_values = flag_values or Thrower()
    392         ql = expr_translate.QL(Thrower(), ThrowException, ThrowException,

RuleCompileException: Annotation may not use variables, but this one uses variable subset.

```

The code can only pass, with an upper case alphabet on the first letter of the variable, like

```

%%logica Subset

@OrderBy(Subset, "cat_id desc");
@Limit(Subset, 10);
Subset(part_num:, cat_id:, name:) :- 
  hao_test.parts(part_num:, name:, part_cat_id: cat_id);
```

**Comment by EvgSkv** [comment]
Thank you for your question! Sorry for delay with the response.

Yes, this is intended behavior. It should definitely be better documented and maybe in the future we'll find a way to do this more intuitive.

The cause is that literals starting with lower case are interpreted as variables, and starting with capital case are interpreted as predicate names. So statement `@Limit(subset, 10);` is interpreted as a limit of undefined variable thus causing error.

An exception is made for predicate calls, i.e. expressions like `subset(part_num:, cat_id:, name:)` where whatever
is called is always interpreted as predicate name. This exception is made to make calls to existing database tables more streamlined.

If for whatever reason you want to keep your predicate lowercased you write the program as:

```
@OrderBy("subset", "cat_id desc");
@Limit("subset", 10);
subset(part_num:, cat_id:, name:) :- 
  hao_test.parts(part_num:, name:, part_cat_id: cat_id);
```

But uppercasing the predicate name, which you did, is the canonical way of writing the code.

Let me know if you have further questions or comments.
